pkg/kubelet/network/dns/dns.go

```diff
@@ -241,19 +247,99 @@ func (c *Configurer) GetPodDNS(pod *v1.Pod) (*runtimeapi.DNSConfig, error) {
+	if dnsPolicy == v1.DNSNone {
+		if utilfeature.DefaultFeatureGate.Enabled(features.CustomPodDNS) {
+			return podDNSNone
+		}
+		// Fallback to DNSClusterFirst. This should not happen as kube-apiserver should
+		// have rejected setting dnsPolicy to DNSNone when feature gate is disabled.
+		useClusterFirstPolicy = true
+	} else {
+		// Only attempt to use ClusterFirst Policy when:
+		// 1. dnsPolicy set to DNSClusterFirst on non-hostnetwork pod, or
+		// 2. dnsPolicy set to DNSClusterFirstWithHostNet.
+		// This behavior is defined by API.
+		useClusterFirstPolicy = ((dnsPolicy == v1.DNSClusterFirst && !kubecontainer.IsHostNetworkPod(pod)) || dnsPolicy == v1.DNSClusterFirstWithHostNet)
+	}
+
+	if useClusterFirstPolicy {
+		if len(c.clusterDNS) != 0 {
+			return podDNSCluster
+		}
 		// clusterDNS is not known.
-		// pod with ClusterDNSFirst Policy cannot be created
-		c.recorder.Eventf(pod, v1.EventTypeWarning, "MissingClusterDNS", "kubelet does not have ClusterDNS IP configured and cannot create Pod using %q policy. Falling back to DNSDefault policy.", pod.Spec.DNSPolicy)
-		log := fmt.Sprintf("kubelet does not have ClusterDNS IP configured and cannot create Pod using %q policy. pod: %q. Falling back to DNSDefault policy.", pod.Spec.DNSPolicy, format.Pod(pod))
+		// Pod with ClusterDNSFirst Policy cannot be created.
+		c.recorder.Eventf(pod, v1.EventTypeWarning, "MissingClusterDNS", "kubelet does not have ClusterDNS IP configured and cannot create Pod using %q policy. Falling back to DNSDefault policy.", dnsPolicy)
+		log := fmt.Sprintf("kubelet does not have ClusterDNS IP configured and cannot create Pod using %q policy. pod: %q. Falling back to DNSDefault policy.", dnsPolicy, format.Pod(pod))
 		c.recorder.Eventf(c.nodeRef, v1.EventTypeWarning, "MissingClusterDNS", log)
+		// Fallback to DNSDefault.
+	}
+	return podDNSHost
+}
+
+// Merge DNS options. If duplicated, entries given by PodDNSConfigOption will
+// overwrite the existing ones.
+func mergeDNSOptions(existingDNSConfigOptions []string, dnsConfigOptions []v1.PodDNSConfigOption) []string {
+	optionsMap := make(map[string]string)
+	for _, op := range existingDNSConfigOptions {
+		var opName, opValue string
+		if index := strings.Index(op, ":"); index != -1 {
+			opName = op[:index]
+			opValue = op[index+1:]
+		} else {
+			opName = op
+		}
+		optionsMap[opName] = opValue
+	}
+	for _, op := range dnsConfigOptions {
+		if op.Value != nil {
+			optionsMap[op.Name] = *op.Value
+		} else {
+			optionsMap[op.Name] = ""
+		}
+	}
+	// Reconvert DNS options into a string array.
+	options := []string{}
+	for opName, opValue := range optionsMap {
+		op := opName
+		if len(opValue) > 0 {
+			op = op + ":" + opValue
+		}
+		options = append(options, op)
+	}
+	return options
+}
+
+// appendDNSConfig appends DNS servers, search paths and options given by
+// PodDNSConfig to the existing DNS config. Duplicated entries will be merged.
+// This assume existingDNSConfig and dnsConfig are not nil.
```

Comment:
assumes,

also, is it necessary to doc this? I would think this is assumed to be the case.

Reply:
Done. I think it doesn't hurt. I passed nil into this function until I wrote the unit test :)